Speed up omnibus build and allow running it dockerized

xvello opened the pull request:
### What does this PR do?

- Create a new `inv docker.dockerize-omnibus` task to run the agent omnibus build in a sandboxed docker container
- Remove the `protobuf-py` dependency, as our UCS4 python build can now use the upstream binary wheel
- ~Remove the `always_build` directive from datadog-process-agent to avoid unnecessary dirtying the cache~
- Blacklist some agent folders that can hold big artifact, to reduce the copy time of the `datadog-agent` recipe
- Reorder the omnibus deps to optimize cache use

Cache (venv, sources and binaries is kept in `~/.omni-build/` by default. We reuse the publicly available circleci runner image for simplicity.

Metrics when running in the `ci_box` on a macbook pro with 4 GB / 4 vcores:
- First run: 42 minutes
- Cache size: 1.9 GB
- Second run: 14 minutes

### Motivation

Make it faster to iterate on the containerized agent, as the docker image depends on the debian package

xvello committed:
Speed up omnibus build and allow running it dockerized

xvello committed:
move integrations-core lower (high change rate)

olivielpeau commented on omnibus/config/projects/agent.rb:
```diff
@@ -115,22 +115,25 @@
+dependency 'datadog-agent-integrations'
+
 if osx?
   dependency 'datadog-agent-mac-app'
```
nit: it'd be nice to keep this one below `datadog-agent` because it depends on `datadog-agent` (omnibus will re-order the definitions anyway, but it's still nice for readability to keep things close to the actual build order)

xvello replied:
You are right, done!

olivielpeau commented on omnibus/config/software/datadog-agent-integrations.rb:
```diff
@@ -8,10 +8,6 @@
 
 name 'datadog-agent-integrations'
 
-dependency 'datadog-pip'
```
unless I'm missing something we should keep the `datadog-pip` dep here

olivielpeau commented on omnibus/config/software/datadog-process-agent.rb:
```diff
@@ -4,7 +4,6 @@
 # Copyright 2018 Datadog, Inc.
 
 name "datadog-process-agent"
-always_build true
```
I wouldn't remove this for non-development builds: unless the `process_agent_version ` changes between 2 builds, I think omnibus will regard the previously downloaded artifact as being up-to-date. So for example, for consecutive builds with a `process_agent_version` set to `master`, as long as the omnibus cache is present I'd expect omnibus to never re-download the process-agent binary even if the binary for `master` on S3 has changed. When we re-enable the omnibus cache on our build pipeline this means that the process-agent won't be updated in non-tagged builds the way it should be, which is an unexpected behavior.

So, overall, I'd advise setting this to `false` only if the user really means that they don't care which version of the process-agent is included in the package.

xvello replied:
The rationale here is to avoid this recipe invalidating the build cache.
Omnibus cache invalidation is pretty naive: if one item has been rebuilt, assume all the following items need to be rebuilt also.

I'll revert this, and investigate an optional "quick-build" switch to skip the systematic rebuild

olivielpeau commented on omnibus/config/software/datadog-agent.rb:
```diff
@@ -16,7 +16,9 @@
-source path: '..'
+source path: '..',
+       # Ignore folders holding big unused artifacts to speed up the copy
+       options: { exclude: ['bin', 'omnibus/pkg', 'Dockerfiles'] }
```
Not sure I follow what we&omnibus are doing here: why do these artifacts end up in the parent directory in the first place?

xvello replied:
Instead of building the agent in place in the given source path, omnibus seems to want to copy everything to a new folder. This is pretty expensive, especially if the sources are in a virtualbox shared folder.

The idea is to speed it up by omitting:
  - `bin/`: that may contain locally-build agent/dogstatsd/cluster-agent binaries
  - `omnibus/pkg`: previous omnibus results (not removed by default, and legit to keep around in some cases)
  - `Dockerfiles/`: may contain agent/dogstatsd/cluster-agent artifacts used when working on the dockerfiles

For the last one, one of the limitation of dockerfiles is that they cannot access any file out of the build folder. This means the debian package cannot be retrieved from `../omnibus/pkg/` but must be copied in this folder beforehand

olivielpeau replied:
I see, thanks for the explanation!

If possible it'd be nice to avoid changing this software definition in such a specific way:
* I'm surprised `bin/` contains anything since I'd expect the binaries to be present in the new folder you mention (i.e. the build folder), instead of the source folder.
* For `omnibus/pkg`, could we play with the omnibus `base_dir` instead so that omnibus writes to a different directory than the agent source's directory?
* For `Dockerfiles/`, I don't see any workaround...

In any case, I don't think it's worth spending too much time on this, I'm fine with keeping this change in the software def as long as the inline comment is a bit more explicit about what the `exclude` option is for, and in which cases it's useful 👍

olivielpeau reviewed (commented):
Left a few comments on the omnibus changes

xvello committed:
Rework omnibus according to feedback:

- move datadog-agent-mac-app back to the end
- re-enable always_build on datadog-process-agent
- datadog-agent-integrations before process to avoid process dirtying the cache
- restore datadog-pip dependency

xvello committed:
revert omnibus changes, keep only invoke task for now
